db/document.go

```diff
@@ -38,6 +39,173 @@ const (
 	DocUnmarshalNone                                     // No unmarshalling (skips import/upgrade check)
 )
 
+// Struct populated by an incoming document
+type IncomingDocument struct {
+	BodyBytes         []byte // BodyBytes contains the raw document only. Doesn't include any special properties
+	Body              Body   // Body may contain the marshalled document only. Doesn't include any special properties until syncFn is generated.
+	SpecialProperties        // Embedded SpecialProperties struct. Extracted either from body or can be manually entered
+}
+
+// Stores special properties which are a part of an incoming document
+type SpecialProperties struct {
+	DocID       string
+	RevID       string
+	Deleted     bool
+	Expiry      uint32
+	Attachments AttachmentsMeta
+	Revisions   Revisions
+}
+
+// Simply constructs IncomingDocument struct with provided body bytes
+func NewIncomingDocument(body []byte) *IncomingDocument {
+	return &IncomingDocument{
+		BodyBytes:         body,
+		SpecialProperties: SpecialProperties{},
+	}
+}
+
+// This must never be ran before GetSyncFnBody
+func (doc *IncomingDocument) GetBody() (map[string]interface{}, error) {
+	if doc.Body == nil {
+		buf := bytes.NewBuffer(doc.BodyBytes)
+		d := base.JSONDecoder(buf)
+		d.UseNumber()
+		err := d.Decode(&doc.Body)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return doc.Body, nil
+}
+
+// Must ensure that the returned body here is never mutated. In the case where we just return body with stamped in
```

Comment:
If the sync function is able to mutate the body then we could definitely run into issues. I was under the impression the user can't update the body via the sync function? Unless we're doing it? 
I'm not sure of a solution if the body can be mutated without copying?

Reply:
Functionally the user can't mutate a document via the sync function BECAUSE we take a deep copy before we run it through the sync fn, and don't reuse the body

My point being was that we need to retain that copy to keep that behaviour. We need to stop users from being able to modify the doc.

Reply:
Right okay, opted to unmarshal, can see what you think